lib/usd/translators/jointWriter.cpp

```diff
@@ -184,9 +184,10 @@ static bool _FindDagPoseMembers(
 
     MPlugArray inputs;
 
-    indices->clear();
-    indices->resize(
-        std::min(membersPlug.numElements(), static_cast<unsigned int>(dagPaths.size())), -1);
+    indices.clear();
+    // TODO: (yliangsiew) Need to know why we even have mismatched array sizes here.
+    indices.resize(
+        std::max(membersPlug.numElements(), static_cast<unsigned int>(dagPaths.size())), -1);
```

Comment:
The way this reads to me, I actually don't think either of these are correct. Based on what the code is attempting to do, I think the intent was for `indices` to have the same size as `dagPaths`:

```C++
indices.resize(dagPaths.size(), -1);
```

We look through the elements of `membersPlug` below and mark the indices of all of the nodes represented there. If there are elements remaining in `dagPaths` that were **not** represented, that ends up being an error and we return false.

Note that `indices` also needs to be a `std::vector` of `int`, not `unsigned int`, for the default value of -1 to be meaningful.

Reply:
I think this has to remain `unsigned int`, because it's used to store
potentially the max. no. of elements in an array plug, which is an `unsigned
int`:
```
for (unsigned int i = 0; i < membersPlug.numElements(); ++i) {
    MPlug memberPlug = membersPlug[i];
    memberPlug.connectedTo(inputs, /*asDst*/ true, /*asSrc*/ false);
    for (unsigned int j = 0; j < inputs.length(); ++j) {
        MObjectHandle connNode(inputs[j].node());
        auto          it = pathIndexMap.find(connNode);
        if (it != pathIndexMap.end()) {
            (*indices)[it->second] = i;
        }
    }
}
```
However, if we can assume that `dagPaths` and `indices` must be the same size,
then the default value of `-1` (which I agree makes zero sense) can now be
dropped. I just did `std::max` as a hacky hotfix previously.

Reply:
Fair enough. I buy that we want to support the maximum possible number of elements in an array plug. I'd argue though that as it stands now, we would hit an error for any elements with indices that exceed the capacity of `int` because of this:
https://github.com/Autodesk/maya-usd/blob/566b7a006f867d07c395ae7911e543a60bda6b00/lib/usd/translators/jointWriter.cpp#L200
We're grabbing an `unsigned int` there and stuffing it in a signed `int`, so any values larger than `std::numeric_limits<int>::max()` are going to wrap around and become negative. The `if (index < 0) {` that follows will then be `true` and we'll bail out in error.
Admittedly, to reach values that trigger that, `membersPlug.numElements()` would need to be up above two billion, so that particular issue is mostly academic. :)

The main issue though is that we need a way to detect whether all of the passed in `dagPaths` were indeed present in the `members` plug. Previously, I think the `indices` vector was trying to do double duty with the `-1` default value and mark which `dagPaths` had been seen so that the final loop could verify that they were all accounted for. We don't have any default value now, so with uninitialized values there's no way to tell if one was missing. But there may also not be an available `unsigned int` value we could use as a sentinel in `indices`. Maybe we just need a separate structure to account for that? Something like this after we do the `indices.resize()`?:
```C++
std::vector<bool> seenPaths(dagPaths.size(), false);
```

Then in the `membersPlug.numElements()` loop, we just mark each path as having been seen when we store an index for it:
```C++
seenPaths[it->second] = true;
```

The validation loop at the end would then just check for any `false` values.

Reply:
Thanks for catching that. You're right; there's still potential for overflow here. I've gone ahead and made the requested changes (i.e. adding a visited array) that should hopefully avoid this issue.